lib/utils/fs.go

```diff
@@ -241,15 +242,64 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
 	return fileLock.Unlock, nil
 }
 
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
+	// Match os.RemoveAll protections in not permitting removal of "." directories
+	if path == "." || (len(path) >= 2 && path[len(path)-1] == '.' && os.IsPathSeparator(path[len(path)-2])) {
+		return &os.PathError{Op: "RemoveAllSecure", Path: path, Err: syscall.EINVAL} // error type matches os.RemoveAll
+	}
+
+	if info, err := os.Lstat(path); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return trace.ConvertSystemError(err)
+	} else if !info.IsDir() {
+		// file or symlink
+		return RemoveSecure(path)
+	}
+	var removeErrors []error
+	if files, err := os.ReadDir(path); err != nil {
+		// don't fail fast, allow removal at end to be attempted
+		removeErrors = []error{err}
+	} else {
+		for _, f := range files {
+			if err := RemoveAllSecure(filepath.Join(path, f.Name())); err != nil {
+				removeErrors = append(removeErrors, err)
+			}
+		}
+	}
+	if err := os.Remove(path); err != nil {
+		removeErrors = append(removeErrors, err)
+	}
+	if len(removeErrors) > 0 {
+		if len(removeErrors) == 1 {
+			return trace.ConvertSystemError(removeErrors[0])
+		}
+		return trace.NewAggregate(removeErrors...)
+	}
+	return nil
+}
+
 // RemoveSecure attempts to securely delete the file by first overwriting the file with random data three times
 // followed by calling os.Remove(filePath).
 func RemoveSecure(filePath string) error {
+	var overwriteErr error
 	for i := 0; i < 3; i++ {
 		if err := overwriteFile(filePath); err != nil {
-			return trace.Wrap(err)
+			if overwriteErr == nil {
+				overwriteErr = err
+			}
 		}
 	}
-	return trace.ConvertSystemError(os.Remove(filePath))
+	// regardless of above errors we want to attempt a removal of the file
+	removeErr := os.Remove(filePath)
+	if overwriteErr == nil {
+		return trace.ConvertSystemError(removeErr)
+	} else if removeErr == nil {
+		return trace.ConvertSystemError(overwriteErr)
+	}
+	return trace.NewAggregate(overwriteErr, removeErr)
```

Comment:
I wonder if we should even bother with an aggregate here. We really only care about the remove error, right? If the overwrite failed but we were still able to remove the file, there's not really any more action that the caller can take to handle the error - the file is gone.

Reply:
I considered this as well.  I wrote it as is because I felt like this was the most complete, but I had considered simply ignoring any overwrite error entirely.  Ultimately if the remove occurred but the overwrite failed for some reason, there is little actionable recourse the user can do.

Let me know your preference, it's fairly corner case either way and I don't feel strongly about it (beyond that I think it's important to try to remove regardless of potential overwrite errors)

Reply:
Don't feel strongly either way and agree it's probably a corner case.

That said, this is primarily intended to be used in user-facing client tools like tsh, and aggregate errors are pretty tricky to understand since they just concatenate two seemingly unrelated error messages. Given that, if I had to choose, I would lean towards just ignoring the overwrite error.

Reply:
Updated, thank you again!